### PR TITLE
Short-circuit Rc/Arc equality checking on equal pointers where T: Eq

### PR DESCRIPTION
**Pull request by chpio**
based on #42965

Is the use of the private trait ok this way? Is there anything else needed for this to get pulled?

### CODE REVIEW

File: src/liballoc/rc.rs [hunk 1]
```diff
@@ -900,12 +901,47 @@ impl<T: Default> Default for Rc<T> {
     }
 }
 
+#[stable(feature = "rust1", since = "1.0.0")]
+trait RcEqIdent<T: ?Sized + PartialEq> {
+    fn eq(&self, other: &Rc<T>) -> bool;
+    fn ne(&self, other: &Rc<T>) -> bool;
```
Comment (oli-obk) [inline src/liballoc/rc.rs]: Couldn't this also have an impl right here that just calls `!self.eq(other)`? That way the impls don't need the trivial implementations given.

**Review by oli-obk — COMMENTED** [review]
Do we have tests anywhere along the lines of

```rust
    let x = std::rc::Rc::new(std::f32::NAN);
    assert_ne!(x, x);
```

File: src/liballoc/rc.rs (hunk 1 above)
Comment (chpio) [reply src/liballoc/rc.rs]: yeah, the current impl is doing this as well (<https://github.com/rust-lang/rust/blob/77a6a61f066af3dd693d8527a8a1bf5a446d295c/src/liballoc/rc.rs#L936-L939>) i don't know why it's not just using the default impl of ne. So i copied that behaviour.

Comment (chpio) [reply src/liballoc/rc.rs]: yes, you'r right

Comment (chpio) [reply src/liballoc/rc.rs]: but what if `T` has a custom impl which isn't `!self.eq(other)`, especially for the default (non-`Eq`) case?

Comment (oli-obk) [reply src/liballoc/rc.rs]: Oh right. Performance is a point here. Indeed, leave it in

File: src/liballoc/rc.rs [hunk 2]
```diff
@@ -8,6 +8,7 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
+#![feature(specialization)]
```
Comment (kennytm) [inline src/liballoc/rc.rs]: This should be placed in `lib.rs`, if it didn't exist already

File: src/liballoc/sync.rs [hunk 3]
```diff
@@ -8,6 +8,7 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
+#![feature(specialization)]
```
Comment (kennytm) [inline src/liballoc/sync.rs]: (remove this)

File: src/liballoc/rc.rs [hunk 4]
```diff
@@ -900,12 +901,47 @@ impl<T: Default> Default for Rc<T> {
...
 
+#[stable(feature = "rust1", since = "1.0.0")]
+trait RcEqIdent<T: ?Sized + PartialEq> {
+    fn eq(&self, other: &Rc<T>) -> bool;
+    fn ne(&self, other: &Rc<T>) -> bool;
+}
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<T: ?Sized + PartialEq> RcEqIdent<T> for Rc<T> {
+    #[inline(always)]
```
Comment (kennytm) [inline src/liballoc/rc.rs]: Do we need `#[inline(always)]` here? Maybe `#[inline]` is enough?

File: src/liballoc/rc.rs (hunk 1 above)
Comment (chpio) [reply src/liballoc/rc.rs]: > Any manual implementation of ne must respect the rule that eq is a strict inverse of ne; that is, !(a == b) if and only if a != b.
>
> -- https://doc.rust-lang.org/std/cmp/trait.PartialEq.html

> Actually there may be good reason for the impls. LLVM is a strange beast sometimes. I certainly wouldn't remove the impls without prior careful measurement.
>
> -- https://github.com/rust-lang/rust-clippy/issues/86

so the whole point of `PartialEq::ne` even being a thing is performance?

**Comment by chpio** [comment]
@oli-obk added a few tests, but it think they are somewhat wanky, testing the specific impl.

**Comment by withoutboats** [comment]
We document that Eq is supposed to be reflexive, but unsafe code cannot rely on that, so its plausible that some implementations aren't following that. Isn't it possible this will break someone's code in which they had a non-reflexive PartialEq that still impl'd Eq? Are we fine with that?

**Comment by SimonSapin** [comment]
Code would only break if it relies on a buggy implementation of `Eq` to stay buggy. I have a hard time to think of a "real" case where that could happen.

**Comment by withoutboats** [comment]
I think I'm inclined to agree, mainly from a pragmatic perspective - its not likely to come up in practice. But I'm uncertain what our general guidelines around backwards compatibility in regard to the "guarantees" that are made about the behavior of traits in std. There's nothing `unsafe` about making an Eq impl that isn't reflexive, and you can't rely on that reflexivity to uphold memory invariants. So where is the line drawn on whose backwards compatibility we are willing to guarantee? Which unenforced contracts of traits are considered "bugs" not to uphold?

For example, members of the team have previously argued that types that implement both `ToString` and `FromStr` should be roundtrippable, would we consider a third party library that doesn't uphold that guarantee similarly buggy and be fine if a change to std broke them? That's what I'm trying to clarify, I guess: we can't actually rely on ecosystem impls upholding untyped contracts, how much allowance is there to implement the traits in a manner different from their documented intent?

**Comment by SimonSapin** [comment]
> There's nothing unsafe about making an Eq impl that isn't reflexive, 

But there’s nothing `unsafe` about changing the result of `Rc<Foo>::eq` either, is there?

**Comment by withoutboats** [comment]
@SimonSapin Definitely not. This is about backwards compatibility.

I bring up unsafe because if the contract were reliable for to avoid UB, breaking that code would clearly be permissible. If Eq were an unsafe trait and we could rely on its reflexivity for memory safety, clearly it would be permissible to break code with a non-reflexive Eq impl, because it failed to uphold the invariants its required to as unsafe code.

But the reflexivity is in this different category of contracts, which only exist in the std documentation and aren't part of the language. How do we draw the line about which of those contracts you need to uphold to get backwards compatibility from std, and which are just contracts that std is guaranteeing to uphold?

**Comment by alexcrichton** [comment]
I think the way I typically approach these kinds of problems is that we strive toward the world we want to be in, for example where `Eq` always represents reflexivity. If, however, along the way we do something that ends up having a large practical impact because that's in fact not true (aka not everything is reflexive) then we backpedal and figure out how to work with that.

That may be a bit cavalier though and certainly isn't as conservative as we could be. In any case one thing I've forgotten is that we could certainly run a crater run for this and just see if there's any impact there initially.

@bors: try

**Comment by alexcrichton** [comment]
@craterbot run start=master#bd47d6825bf4090517549d33cfef10d3300b4a75 end=try#aa49d8ef14939ddec0e34b346b60174a5673d48f mode=build-and-test

**Comment by Kimundi** [comment]
I see it similar to the ability to define a `Clone` impl for a `T: Copy` that does more than just a `memcpy`: Its safe to do so, but a violation of the interface, so if some code does not end up calling the method its because you misused the interface.

**Comment by chpio** [comment]
Is crater calling more than just `cargo check`? It would be pretty useless for this PR if it's the case.

**Comment by memoryruins** [comment]
The current agent is in [`build-and-test`](https://github.com/rust-lang-nursery/crater/blob/master/docs/bot-usage.md#available-experiment-modes) mode (check the final portion of the command issued to craterbot).

**Comment by chpio** [comment]
couldn't spot any meaningful regressions here, just some timing or env-var not set failures. is anyone else willing to look into it?

**Comment by alexcrichton** [comment]
I took a look myself and some of them looked suspicious but they were all spurious I think. Many of the failures do indeed use `Arc` and `Rc` internally, but they didn't seem to be failing due to this PR. In that case this looks good to go, thanks again @chpio!

@bors: r+

File: src/liballoc/rc.rs [hunk 5]
```diff
@@ -900,12 +900,47 @@ impl<T: Default> Default for Rc<T> {
...
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<T: ?Sized + Eq> RcEqIdent<T> for Rc<T> {
+    #[inline]
+    fn eq(&self, other: &Rc<T>) -> bool {
+        Rc::ptr_eq(self, other) || **self == **other
+    }
+
+    #[inline]
+    fn ne(&self, other: &Rc<T>) -> bool {
+        !Rc::ptr_eq(self, other) && **self != **other
```
Comment (RalfJung) [inline src/liballoc/rc.rs]: It might be more readable to use `!RcEqIdent::eq` here.

**Comment by RalfJung** [comment]
Wouldn't it make sense to also do the same when comparing `&[mut] T`? That seems to be the far more obvious case than `Rc`/`Arc`.

And then comparison of `Rc`/`Arc` could be implemented via `deref`.

File: src/liballoc/rc.rs (hunk 5 above)
Comment (chpio) [reply src/liballoc/rc.rs]: https://github.com/rust-lang/rust/pull/56550#discussion_r239363341 ?

**Comment by chpio** [comment]
@RalfJung https://github.com/rust-lang/rust/pull/42965#issuecomment-312702479

**Comment by chpio** [comment]
> I think such conclusions should be added as comments in the source code, they'll be lost otherwise.

hmm, yeah. are you saying i should make a new pr for that?

**Comment by RalfJung** [comment]
That would be great :)

**Pull request merged**